connector/profilingmetricsconnector/frame.go

```diff
@@ -123,67 +190,161 @@ func fetchFrameInfo(dictionary pprofile.ProfilesDictionary,
+	}
 }
 
-// classifyFrame classifies sample into one or more categories based on frame type.
+// classifyFrames classifies a sample into one or more categories based on frame information.
 // This takes place by incrementing the associated metric count.
-func classifyFrame(dictionary pprofile.ProfilesDictionary,
+func (c *profilesToMetricsConnector) classifyFrames(dictionary pprofile.ProfilesDictionary,
 	locationIndices pcommon.Int32Slice,
-	sample pprofile.Sample,
 	counts map[metric]int64,
 	nativeCounts map[string]int64,
+	kernelCounts map[kernelStackAttr]int64,
+	multiplier int64,
 ) error {
-	fi, err := fetchFrameInfo(dictionary, locationIndices, 0)
-	if err != nil {
-		return err
-	}
+	var err error
+
+	className := ""
+	syscall := ""
+	lastMatchIdx := len(classes)
+	hasKernel := false
+	for idx := range locationIndices.Len() {
+		fi, err := c.fetchFrameInfo(dictionary, locationIndices, idx)
+		if err != nil {
+			break
+		}
 
-	leafFrameType := fi.typ
-	// We don't need a separate metric for total number of samples, as this can always be
-	// derived from summing the metricKernel and metricUser counts.
-	metric := allowedFrameTypes[leafFrameType]
+		if idx == 0 {
+			// Only need to call this once per stacktrace
+			classifyLeaf(fi, counts, nativeCounts, multiplier)
+		}
+
+		// Kernel frame specific logic follows
+		if fi.typ != frameTypeKernel {
+			break
+		}
 
-	// TODO: Scale all counts by number of events in each Sample. Currently,
-	// this logic assumes 1 event per Sample (thus the increments by 1 below),
-	// which isn't necessarily the case.
+		hasKernel = true
+		if fi.functionName == "" {
+			// No classification possible for this frame
+			continue
+		}
+
+		for cIdx, class := range classes {
+			// Priority matching logic
+			if cIdx == lastMatchIdx {
+				// Early exit if there's no higher priority match possible
+				break
+			}
+			if class.rx.MatchString(fi.functionName) {
+				className = class.name
+				lastMatchIdx = cIdx
+				break
+			}
+		}
 
-	if leafFrameType != frameTypeKernel {
-		counts[metricUser]++
+		// Try to match a syscall and avoid string allocations by using indices
+		// to string location.
+		sysIndices := syscallRx.FindStringSubmatchIndex(fi.functionName)
+		if len(sysIndices) == 4 {
+			syscall = fi.functionName[sysIndices[2]:sysIndices[3]]
+			// If we match a syscall, we don't need to keep iterating frames
+			// as there are no further (syscall or category) matches possible.
+			break
+		}
 	}
 
-	if leafFrameType != frameTypeNative {
-		counts[metric]++
-		return nil
+	if !hasKernel {
+		return err
 	}
 
-	// Extract native library name and increment associated count
-	if sm := rx.FindStringSubmatch(fi.filename); sm != nil {
-		nativeCounts[sm[1]]++
+	if className == "" && syscall == "" {
+		// No kernel category / syscall match
+		counts[metricKernel] += multiplier
 	} else {
-		counts[metric]++
+		kernelCounts[kernelStackAttr{
+			class:   className,
+			syscall: syscall}] += multiplier
```

Comment:
If found this a bit misleading, shouldn't `counts[metricKernel]` always be incremented? Even if we did not match the category/syscall, we are sure it is from the kernel

Reply:
Thanks for the clarifications, it helped :+1: